rust/kernel/spi.rs

```diff
@@ -0,0 +1,279 @@
+/// kernel's default implementations will be used.
+pub const USE_NONE: ToUse = ToUse {
+    probe: false,
+    remove: false,
+    shutdown: false,
+};
+
+/// Corresponds to the kernel's spi_driver's methods. Implement this trait on a type to
+/// express the need of a custom probe, remove or shutdown function for your SPI driver.
+/// Use the [`declare_spi_methods`] macro to declare which methods you wish to use.
+pub trait SpiMethods {
+    /// The methods to define. Use [`declare_spi_methods`] to declare them
+    const TO_USE: ToUse;
+
+    /// Corresponds to the kernel's spi_driver's probe method field
+    fn probe(mut _spi_dev: SpiDevice) -> Result {
+        Ok(())
+    }
+
+    /// Corresponds to the kernel's spi_driver's remove method field
+    fn remove(mut _spi_dev: SpiDevice) -> Result {
+        Ok(())
+    }
+
+    /// Corresponds to the kernel's spi_driver's shutdown method field
+    fn shutdown(mut _spi_dev: SpiDevice) {}
+}
+
+/// Populate the TO_USE field in the [`SpiMethods`] implementer
+///
+/// ```rust
+/// impl SpiMethods for MySpiMethods {
+///     /// Let's say you only want a probe and remove method, no shutdown
+///     declare_spi_methods!(probe, remove);
+///
+///     /// Define your probe and remove methods. If you don't, default implementations
+///     /// will be used instead. These default implementations do NOT correspond to the
+///     /// kernel's default implementations! If you wish to use the Kernel's default
+///     /// spi functions implementations, do not declare them using the `declare_spi_methods`
+///     /// macro. For example, here our Driver will use the Kernel's shutdown method.
+///     fn probe(spi_dev: SpiDevice) -> Result {
+///         // ...
+///
+///         Ok(())
+///     }
+///
+///     fn remove(spi_dev: SpiDevice) -> Result {
+///         // ...
+///
+///         Ok(())
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! declare_spi_methods {
+    () => {
+        const TO_USE: $crate::spi::ToUse = $crate::spi::USE_NONE;
+    };
+    ($($method:ident),+) => {
+        const TO_USE: $crate::spi::ToUse = $crate::spi::ToUse {
+            $($method: true),+,
+            ..$crate::spi::USE_NONE
+        };
+    };
+}
+
+impl DriverRegistration {
+    fn new(this_module: &'static crate::ThisModule, name: &'static CStr) -> Self {
+        DriverRegistration {
+            this_module,
+            name,
+            registered: false,
+            spi_driver: bindings::spi_driver::default(),
+        }
+    }
+
+    /// Create a new `DriverRegistration` and register it. This is equivalent to creating
+    /// a static `spi_driver` and then calling `spi_driver_register` on it in C.
+    ///
+    /// ```
+    /// // The function returns a result, which you need to handle. Use the `?` operator
+    /// // to propagate it
+    /// let spi_driver =
+    ///     spi::DriverRegistration::new_pinned::<MySpiMethods>(&THIS_MODULE, cstr!("my_driver_name"))?;
+    /// ```
```

Comment:
Couple suggestions:
  - Add `# Examples` section.
  - I think explaining `Result` and `?` should not be done here -- it seems something people should already know.

```suggestion
    /// Create a new `DriverRegistration` and register it. This is equivalent to creating
    /// a static `spi_driver` and then calling `spi_driver_register` on it in C.
    ///
    /// # Examples
    ///
    /// ```
    /// let spi_driver =
    ///     spi::DriverRegistration::new_pinned::<MySpiMethods>(&THIS_MODULE, cstr!("my_driver_name"))?;
    /// ```
```

Reply:
Fixed by 92a19c2ef2d0d0780c34327b273a21db627af711